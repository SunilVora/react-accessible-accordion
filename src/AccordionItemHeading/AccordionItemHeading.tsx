import { default as classnames } from 'classnames';
import * as React from 'react';
import { UUID } from '../ItemContainer/ItemContainer';

type AccordionItemHeadingProps = React.HTMLAttributes<HTMLDivElement> & {
    hideBodyClassName: string;
    expanded: boolean;
    uuid: UUID;
    disabled: boolean;
    focus: boolean;
    removeFocus(uuid: UUID): void;
    setExpanded(uuid: UUID, expanded: boolean): void;
    setFocusToHead(): void;
    setFocusToTail(): void;
    setFocusToPrevious(uuid: UUID): void;
    setFocusToNext(uuid: UUID): void;
};

type AccordionItemHeadingState = {};

export default class AccordionItemHeading extends React.Component<
    AccordionItemHeadingProps,
    AccordionItemHeadingState
> {
    focusRef: React.RefObject<HTMLDivElement> = React.createRef();

    componentDidUpdate(): void {
        if (this.props.focus) {
            this.focusRef.current.focus();
        }
    }

    handleClick = (): void => {
        const { uuid, expanded, setExpanded } = this.props;

        setExpanded(uuid, !expanded);
    };

    handleKeyPress = (evt: React.KeyboardEvent<HTMLDivElement>): void => {
        if (evt.charCode === 13 || evt.charCode === 32) {
            evt.preventDefault();
            this.handleClick();
        }
    };

    handleKeyDown = (evt: React.KeyboardEvent<HTMLDivElement>): void => {
        const {
            uuid,
            setFocusToHead,
            setFocusToTail,
            setFocusToPrevious,
            setFocusToNext,
        } = this.props;

        switch (evt.which) {
            case 35:
                evt.preventDefault();
                setFocusToTail();
                break;
            case 36:
                evt.preventDefault();
                setFocusToHead();
                break;
            case 38:
                evt.preventDefault();
                setFocusToPrevious(uuid);
                break;
            case 40:
                evt.preventDefault();
                setFocusToNext(uuid);
                break;
            default:
                evt.preventDefault();
        }
    };

    handleBlur = (): void => {
        this.props.removeFocus(this.props.uuid);
    };

    render(): JSX.Element {
        const {
            className,
            hideBodyClassName,
            setExpanded,
            expanded,
            uuid,
            disabled,
            focus,
            setFocusToHead,
            setFocusToTail,
            setFocusToPrevious,
            setFocusToNext,
            removeFocus,
            ...rest
        } = this.props;

        const id = `accordion__heading-${uuid}`;
        const ariaControls = `accordion__panel-${uuid}`;
        const role = 'button';
        const titleClassName = classnames(className, {
            [hideBodyClassName]: hideBodyClassName && !expanded,
        });
        const onClick = disabled ? undefined : this.handleClick;

        return (
            <div
                id={id}
                ref={this.focusRef}
                aria-expanded={expanded}
                aria-controls={ariaControls}
                className={titleClassName}
                onClick={onClick}
                role={role}
                tabIndex={0}
                onKeyPress={this.handleKeyPress}
                onKeyDown={this.handleKeyDown}
                onBlur={this.handleBlur}
                {...rest}
            />
        );
    }
}
