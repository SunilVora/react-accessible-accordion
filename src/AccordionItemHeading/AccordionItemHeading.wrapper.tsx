import * as React from 'react';
import {
    CONTEXT_KEY as ACCORDION_CONTEXT_KEY,
    getAccordionStore,
    Item,
} from '../AccordionContainer/AccordionContainer';
import * as propTypes from '../helpers/propTypes';
import {
    CONTEXT_KEY as ITEM_CONTEXT_KEY,
    getItemStore,
} from '../ItemContainer/ItemContainer';
import { default as AccordionItemHeading } from './AccordionItemHeading';

type AccordionItemHeadingWrapperProps = React.HTMLAttributes<HTMLDivElement> & {
    hideBodyClassName: string;
};

type AccordionItemHeadingWrapperState = {};

type AccordionItemHeadingWrapperContext = {
    [ACCORDION_CONTEXT_KEY](): null;
    [ITEM_CONTEXT_KEY](): null;
};

export default class AccordionItemHeadingWrapper extends React.Component<
    AccordionItemHeadingWrapperProps
> {
    static contextTypes: AccordionItemHeadingWrapperContext = {
        [ACCORDION_CONTEXT_KEY]: propTypes.wildcard,
        [ITEM_CONTEXT_KEY]: propTypes.wildcard,
    };

    static defaultProps: AccordionItemHeadingWrapperProps = {
        className: 'accordion__heading',
        hideBodyClassName: '',
    };

    render(): JSX.Element {
        const accordionStore = getAccordionStore(this.context);

        if (!accordionStore) {
            // tslint:disable-next-line:no-console
            console.error(
                'AccordionItemHeading component cannot render because it has not been nested inside an Accordion component.',
            );

            return null;
        }

        const itemStore = getItemStore(this.context);

        if (!itemStore) {
            // tslint:disable-next-line:no-console
            console.error(
                'AccordionItemHeading component cannot render because it has not been nested inside an AccordionItem component.',
            );

            return null;
        }

        const { uuid } = itemStore;
        const { items } = accordionStore;
        const item = items.filter(
            (stateItem: Item) => stateItem.uuid === uuid,
        )[0];

        return (
            <AccordionItemHeading
                {...this.props}
                {...item}
                setExpanded={accordionStore.setExpanded}
                setFocusToHead={accordionStore.setFocusToHead}
                setFocusToTail={accordionStore.setFocusToTail}
                setFocusToPrevious={accordionStore.setFocusToPrevious}
                setFocusToNext={accordionStore.setFocusToNext}
                removeFocus={accordionStore.removeFocus}
            />
        );
    }
}
