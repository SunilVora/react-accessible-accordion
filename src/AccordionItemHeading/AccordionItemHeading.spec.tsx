import { mount, ReactWrapper } from 'enzyme';
import * as React from 'react';
import {
    Item,
    Provider as AccordionProvider,
} from '../AccordionContainer/AccordionContainer';
import { Provider as ItemProvider } from '../ItemContainer/ItemContainer';
import { default as AccordionItemHeading } from './AccordionItemHeading.wrapper';

describe('AccordionItemHeading', () => {
    const DEFAULT_ITEM: Item = {
        uuid: 0,
        expanded: false,
        disabled: false,
        focus: false,
    };

    it('renders null outside the context of an ‘Accordion’', () => {
        const wrapper = mount(
            <ItemProvider uuid="foo">
                <AccordionItemHeading />
            </ItemProvider>,
        );

        expect(wrapper.html()).toBeNull();
    });

    it('renders null outside the context of an ‘AccordionItem’', () => {
        const wrapper = mount(
            <AccordionProvider items={[DEFAULT_ITEM]}>
                <AccordionItemHeading />
            </AccordionProvider>,
        );

        expect(wrapper.html()).toBeNull();
    });

    function mountItem(
        node: React.ReactNode,
        item: Item = DEFAULT_ITEM,
    ): ReactWrapper {
        return mount(
            <AccordionProvider allowMultipleExpanded={true} items={[item]}>
                <ItemProvider uuid={item.uuid}>{node}</ItemProvider>
            </AccordionProvider>,
        );
    }

    function mountItems(
        node: React.ReactNode,
        items: Item[] = [DEFAULT_ITEM],
    ): ReactWrapper {
        return mount(
            <AccordionProvider allowMultipleExpanded={true} items={items}>
                {items.map((item: Item, index: number) => (
                    <ItemProvider key={index} uuid={item.uuid}>
                        {node}
                    </ItemProvider>
                ))}
            </AccordionProvider>,
        );
    }

    function isExpanded(wrapper: ReactWrapper, uuid: string | number): boolean {
        const instance = wrapper
            .find(AccordionProvider)
            .instance() as AccordionProvider;

        return !!instance.state.items.find((item: Item) => item.uuid === uuid)
            .expanded;
    }

    function isFocused(wrapper: ReactWrapper, uuid: string | number): boolean {
        const instance = wrapper
            .find(AccordionProvider)
            .instance() as AccordionProvider;

        return !!instance.state.items.find((item: Item) => item.uuid === uuid)
            .focus;
    }

    function findFocusedIndex(wrapper: ReactWrapper): number {
        const instance = wrapper
            .find(AccordionProvider)
            .instance() as AccordionProvider;

        return instance.state.items.findIndex((item: Item) => item.focus);
    }

    it('renders correctly with min params', () => {
        const wrapper = mountItem(
            <AccordionItemHeading>
                <div>Fake Title</div>
            </AccordionItemHeading>,
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with different className', () => {
        const className = 'className';
        const wrapper = mountItem(
            <AccordionItemHeading className={className} />,
        );
        expect(wrapper.find('div').hasClass(className)).toEqual(true);
    });

    it('renders with different hideBodyClassName', () => {
        const hideBodyClassName = 'hideBodyClassName';
        const wrapper = mountItem(
            <AccordionItemHeading hideBodyClassName={hideBodyClassName} />,
        );
        expect(wrapper.find('div').hasClass(hideBodyClassName)).toEqual(true);
    });

    it("doesn't present hideBodyClassName when collapsed", () => {
        const hideBodyClassName = 'hideBodyClassName';
        const wrapper = mountItem(
            <AccordionItemHeading hideBodyClassName={hideBodyClassName} />,
        );
        expect(wrapper.find('div').hasClass(hideBodyClassName)).toEqual(true);
    });

    it('toggles state when clicked', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('click');
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeTruthy();
    });

    it('doesn’t toggle state when trying to click but disabled', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
            {
                ...DEFAULT_ITEM,
                disabled: true,
            },
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('click');
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
    });

    it('toggles state when pressing enter', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('keypress', { charCode: 13 });
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeTruthy();
    });

    it('toggles state when pressing space', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('keypress', { charCode: 32 });
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeTruthy();
    });

    it('doesn’t toggle state when pressing another key', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('keypress', { charCode: 35 });
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
    });

    it('respects arbitrary user-defined props', () => {
        const wrapper = mountItem(
            <AccordionItemHeading lang="en">Fake Title</AccordionItemHeading>,
        );

        const div = wrapper.find('div').getDOMNode();

        expect(div.getAttribute('lang')).toEqual('en');
    });

    // edge case to cover branch
    it('doesn’t toggle state when clicking but disabled & allowMultipleExpanded === false', async () => {
        const wrapper = mount(
            <AccordionProvider
                items={[
                    {
                        ...DEFAULT_ITEM,
                        disabled: true,
                    },
                ]}
            >
                <ItemProvider uuid={DEFAULT_ITEM.uuid}>
                    <AccordionItemHeading>Fake Title</AccordionItemHeading>
                </ItemProvider>
            </AccordionProvider>,
        );

        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
        wrapper.find('div').simulate('click');
        expect(isExpanded(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
    });

    it('toggles state when blurred', async () => {
        const wrapper = mountItem(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
            {
                ...DEFAULT_ITEM,
                focus: true,
            },
        );

        expect(isFocused(wrapper, DEFAULT_ITEM.uuid)).toBeTruthy();
        wrapper.find('div').simulate('blur');
        expect(isFocused(wrapper, DEFAULT_ITEM.uuid)).toBeFalsy();
    });

    it('sets focus to last item when end is pressed', async () => {
        const wrapper = mountItems(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
            [
                {
                    ...DEFAULT_ITEM,
                    uuid: 'foo',
                    focus: true,
                },
                {
                    ...DEFAULT_ITEM,
                    uuid: 'bar',
                    focus: false,
                },
            ],
        );

        expect(findFocusedIndex(wrapper)).toBe(0);
        wrapper
            .find('div')
            .first()
            .simulate('keydown', { which: 35 });
        expect(findFocusedIndex(wrapper)).toBe(1);
    });

    it('sets focus to first item when home is pressed', async () => {
        const wrapper = mountItems(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
            [
                {
                    ...DEFAULT_ITEM,
                    uuid: 'foo',
                    focus: false,
                },
                {
                    ...DEFAULT_ITEM,
                    uuid: 'bar',
                    focus: true,
                },
            ],
        );

        expect(findFocusedIndex(wrapper)).toBe(1);
        wrapper
            .find('div')
            .first()
            .simulate('keydown', { which: 36 });
        expect(findFocusedIndex(wrapper)).toBe(0);
    });

    it('sets focus to previous item when up arrow is pressed', async () => {
        const wrapper = mountItems(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
            [
                {
                    ...DEFAULT_ITEM,
                    uuid: 'foo',
                    focus: false,
                },
                {
                    ...DEFAULT_ITEM,
                    uuid: 'bar',
                    focus: true,
                },
            ],
        );

        expect(findFocusedIndex(wrapper)).toBe(1);
        wrapper
            .find('div')
            .last()
            .simulate('keydown', { which: 38 });
        expect(findFocusedIndex(wrapper)).toBe(0);
    });

    it('sets focus to next item when down arrow is pressed', async () => {
        const wrapper = mountItems(
            <AccordionItemHeading>Fake Title</AccordionItemHeading>,
            [
                {
                    ...DEFAULT_ITEM,
                    uuid: 'foo',
                    focus: true,
                },
                {
                    ...DEFAULT_ITEM,
                    uuid: 'bar',
                    focus: false,
                },
            ],
        );

        expect(findFocusedIndex(wrapper)).toBe(0);
        wrapper
            .find('div')
            .first()
            .simulate('keydown', { which: 40 });
        expect(findFocusedIndex(wrapper)).toBe(1);
    });
});
