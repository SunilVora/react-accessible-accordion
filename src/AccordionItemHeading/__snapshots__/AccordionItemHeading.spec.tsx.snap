// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItemHeading renders correctly with min params 1`] = `
<div
  aria-controls="accordion__panel-0"
  aria-expanded={false}
  className="accordion__heading"
  id="accordion__heading-0"
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div>
    Fake Title
  </div>
</div>
`;
